crates/context_server/src/transport/sse_transport.rs

```diff
@@ -0,0 +1,143 @@
+
+pub struct SseTransport {
+    message_url: MessageUrl,
+    stdin_receiver: channel::Receiver<String>,
+    stderr_receiver: channel::Receiver<String>,
+    http_client: Arc<dyn HttpClient>,
+}
+
+impl SseTransport {
+    pub fn new(endpoint: Url, cx: &AsyncApp) -> Result<Self> {
+        let (stdin_sender, stdin_receiver) = channel::unbounded::<String>();
+        let (_stderr_sender, stderr_receiver) = channel::unbounded::<String>();
+        let (message_url, url_sender) = MessageUrl::new();
+        let http_client = cx.update(|cx| cx.http_client().clone())?;
+
+        let message_url_clone = message_url.url.clone();
+        cx.spawn({
+            let http_client = http_client.clone();
+            move |cx| async move {
+                Self::handle_sse_stream(
+                    cx.background_executor(),
+                    endpoint,
+                    message_url_clone,
+                    stdin_sender,
+                    url_sender,
+                    http_client,
+                )
+                .await
+                .log_err()
+            }
+        })
+        .detach();
+
+        Ok(Self {
+            message_url,
+            stdin_receiver,
+            stderr_receiver,
+            http_client,
+        })
+    }
+
+    async fn handle_sse_stream(
+        executor: &BackgroundExecutor,
+        endpoint: Url,
+        message_url: Arc<Mutex<Option<String>>>,
+        stdin_sender: channel::Sender<String>,
+        url_sender: channel::Sender<()>,
+        http_client: Arc<dyn HttpClient>,
+    ) -> Result<()> {
+        loop {
+            let mut response = http_client
+                .get(endpoint.as_str(), Default::default(), true)
+                .await?;
+            let mut reader = BufReader::new(response.body_mut());
+            let mut line = String::new();
+
+            loop {
+                futures::select! {
+                    result = reader.read_line(&mut line).fuse() => {
+                        match result {
+                            Ok(n) if n == 0 => break,
+                            Ok(_) => {
+                                if line.starts_with("data: ") {
+                                    let data = line.trim_start_matches("data: ");
+                                    if data.starts_with("http") {
+                                        *message_url.lock().await = Some(data.trim().to_string());
+                                        url_sender.send(()).await?;
+                                    } else {
+                                        stdin_sender.send(data.to_string()).await?;
+                                    }
+                                }
+                                line.clear();
+                            },
+                            Err(_) => break,
+                        }
+                    },
+                    _ = executor.timer(Duration::from_secs(30)).fuse() => {
+                        break;
+                    }
```

Comment:
Here I'm trying to account for computer sleep/wake-up, but it doesn't always work as expected. Is there a better way to handle these interruptions?